File: module-08-input/appendix/test_reference.R
# Exactly three named tests; run with Rscript /absolute/path/test_reference.R.
file_arg <- grep('^--file=', commandArgs(FALSE), value = TRUE)
if (!length(file_arg)) stop('Use Rscript or the documented WebR runner')
here <- dirname(normalizePath(sub('^--file=', '', file_arg[[1]])))
source(file.path(here, 'reference.R'))
cases <- load_cases()
assert_equal <- function(actual, expected) {
  stopifnot(isTRUE(all.equal(actual, expected, check.attributes = FALSE)))
}
check_values <- function(rows) {
  actual <- translate(rows[, source_columns, drop = FALSE])
  assert_equal(actual$num_value, rows$expected_num_value)
  assert_equal(actual$date_value, rows$expected_date_value)
}
test_T1_original_fixture <- function() {
  check_values(cases[cases$case_id == 'fixture', , drop = FALSE])
}
test_T2_valid_leap_day_and_numeric_boundary <- function() {
  check_values(cases[cases$case_id == 'edge', , drop = FALSE])
}
test_T3_source_order_and_result_types <- function() {
  input <- cases[rev(seq_len(nrow(cases))), source_columns, drop = FALSE]
  before <- input
  actual <- translate(input)
  stopifnot(identical(input, before), identical(actual[source_columns], before))
  stopifnot(is.numeric(actual$num_value), inherits(actual$date_value, 'Date'))
  stopifnot(is.character(actual$char_num), is.character(actual$char_date))
  stopifnot('0012' %in% actual$char_num)
}

test_T1_original_fixture(); cat("PASS T1 original_fixture\n")
test_T2_valid_leap_day_and_numeric_boundary(); cat("PASS T2 valid_leap_day_and_numeric_boundary\n")
test_T3_source_order_and_result_types(); cat("PASS T3 source_order_and_result_types\n")

File: module-08-input/appendix/test_reference.py
"""Exactly three acceptance tests; added valid cases do not change trial scores."""
import importlib.util
from pathlib import Path
import unittest
import pandas as pd

spec = importlib.util.spec_from_file_location('input_reference', Path(__file__).with_name('reference.py'))
ref = importlib.util.module_from_spec(spec)
spec.loader.exec_module(ref)

class ReferenceTests(unittest.TestCase):
    def check_values(self, rows):
        actual = ref.translate(rows[ref.SOURCE_COLUMNS])
        self.assertEqual(actual['num_value'].tolist(), rows['expected_num_value'].tolist())
        self.assertEqual(actual['date_value'].dt.strftime('%Y-%m-%d').tolist(), rows['expected_date_value'].tolist())

    def test_T1_original_fixture(self):
        rows = ref.load_cases()
        self.check_values(rows.loc[rows.case_id.eq('fixture')])

    def test_T2_valid_leap_day_and_numeric_boundary(self):
        rows = ref.load_cases()
        self.check_values(rows.loc[rows.case_id.eq('edge')])

    def test_T3_source_order_and_result_types(self):
        source = ref.load_cases()[ref.SOURCE_COLUMNS].iloc[::-1].copy(deep=True)
        before = source.copy(deep=True)
        actual = ref.translate(source)
        self.assertIsNot(actual, source)
        pd.testing.assert_frame_equal(source, before)
        pd.testing.assert_frame_equal(actual[ref.SOURCE_COLUMNS], before)
        self.assertTrue(pd.api.types.is_numeric_dtype(actual['num_value']))
        self.assertTrue(pd.api.types.is_datetime64_any_dtype(actual['date_value']))
        self.assertTrue(all(isinstance(v, str) for v in actual['char_num']))
        self.assertTrue(all(isinstance(v, str) for v in actual['char_date']))
        self.assertIn('0012', actual['char_num'].tolist())

if __name__ == '__main__':
    unittest.main(verbosity=2)
